### PR TITLE
Rustdoc: Do not list impl when trait has doc(hidden)

### LINKED ISSUE #86448: cross-crate `#[doc(hidden)]` is ignored when listing trait implementations
### Context

https://users.rust-lang.org/t/hidden-trait-implementation-in-online-docs/61256/4

### Repro

```bash
set -euo pipefail
cd $(mktemp -d)

cargo new -q --lib dependency
cat> dependency/src/lib.rs <<'EOF'
    #[doc(hidden)]
    pub enum HiddenType {}
    
    pub enum MyLibType {}
    impl From<HiddenType> for MyLibType {
        fn from (it: HiddenType)
          -> MyLibType
        {
            match it {}
        }
    }
EOF
(cd dependency
    set -x
    cargo doc -q --no-deps
    # No matches
    ! grep -o HiddenType target/doc/dependency/index.html
)

cargo init -q --lib --name mylib
cat> src/lib.rs <<'EOF'
    pub use ::dependency::HiddenType; // OK, not re-exported
    
    pub enum MyLibType {}
    impl From<HiddenType> for MyLibType {
        fn from (it: HiddenType)
          -> MyLibType
        {
            match it {}
        }
    }
EOF

cat>> Cargo.toml <<'EOF'
    dependency.path = 'dependency'
EOF

set -x
cargo doc -q --no-deps
# No matches
! grep -o HiddenType target/doc/mylib/index.html
# There are matches
grep -o HiddenType target/doc/mylib/enum.MyLibType.html
```

  - Last line yields several occurrences of `HiddenType` in the generated `.html` file.

    ## Expected behavior

    No occurrences of `HiddenType` anywhere since it was `#[doc(hidden)]`

  - [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=14b5e1c16ef0dc2df505fba4d4e2065c)

@rustbot label C-bug T-rustdoc

cc @GuillaumeGomez

### PR DESCRIPTION
**Pull request by fee1-dead**
Fixes #86448.

### CODE REVIEW

**Comment by fee1-dead** [comment]
Looks like highfive bot skipped this, so r? @jyn514

**Comment by jyn514** [comment]
@danielhenrymantilla are you interested in reviewing this?

**Comment by danielhenrymantilla** [comment]
Yeah I can take care of that, and ping you when I think it's okay 🙂

**Comment by danielhenrymantilla** [comment]
Nice, I like the manually laid out stack to avoid recursing 🙂 A few comments:

  - I've seen you're "only" checking for the generic parameters of a `#[doc(hidden)]` trait, so the following questions remain open:

      - What happens if the cross-crate `#[doc(hidden)]` type is the "receiver" / `Self` of the `impl Trait…`?

        Screen Shot 2021-06-21 at 23 54 36

          - [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=66a24bbd63f1c70766da82160f720499)

          - Given your algorithm, I guess a very simple solution would be to add the `Self` type to that queue of yours 👍 

      - What happens if the _trait_ itself is the cross-crate `#[doc(hidden)]` one?

        Screen Shot 2021-06-22 at 00 03 32

Could you try handling those cases as well? Thanks

File: src/librustdoc/clean/inline.rs [hunk 1]
```diff
@@ -420,6 +423,20 @@ crate fn build_impl(
...
+    while let Some(ty) = deque.pop_back() {
+        if let Some(did) = ty.def_id() {
+            if cx.tcx.get_attrs(did).lists(sym::doc).has_word(sym::hidden) {
+                return;
+            }
+        }
+        if let Some(generics) = ty.generics() {
+            deque.extend(generics);
+        }
+    }
```
Comment (danielhenrymantilla) [inline src/librustdoc/clean/inline.rs:438]: Nit: I don't really see `VecDeque`'s API being used, here, so let's keep it simple and use a `Vec` 🙂 

```suggestion
    let mut stack: Vec<&Type> = trait_.iter().collect();
    while let Some(ty) = stack.pop() {
        if let Some(did) = ty.def_id() {
            if cx.tcx.get_attrs(did).lists(sym::doc).has_word(sym::hidden) {
                return;
            }
        }
        if let Some(generics) = ty.generics() {
            stack.extend(generics);
        }
    }
```

Comment (fee1-dead) [reply src/librustdoc/clean/inline.rs]: Done

**Review by danielhenrymantilla — APPROVED** [review]
Nice, LGTM ✅ (cc @jyn514)

File: src/test/rustdoc/cross-crate-hidden.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Issue #86448: test for cross-crate `doc(hidden)`
+#![crate_name = "foo"]
+
+// aux-build:cross-crate-hidden.rs
+extern crate cross_crate_hidden;
+
+pub use ::cross_crate_hidden::{HiddenType, HiddenTrait}; // OK, not re-exported
```
Comment (jyn514) [inline src/test/rustdoc/cross-crate-hidden.rs:7]: Can you also test this for the same-crate case?
```suggestion
// revisions: extern local
// [extern] aux-build:cross-crate-hidden.rs
#[cfg(extern)]
extern crate cross_crate_hidden;
#[cfg(extern)]
pub use ::cross_crate_hidden::{HiddenType, HiddenTrait}; // OK, not re-exported

#[cfg(local)]
#[doc(hidden)]
pub enum HiddenType {}

#[cfg(local)]
#[doc(hidden)]
pub trait HiddenTrait {}
```
(not sure if revisions work for rustdoc tests - if not, it's fine to make these two separate tests.)

Please also change the file name from `cross-crate-hidden` to `cross-crate-hidden-impl-parameter`, which I think is a more accurate description.

**Review by jyn514 — APPROVED** [review]
This is great, thanks! r=me with the new test added.

File: src/test/rustdoc/cross-crate-hidden.rs (hunk 2 above)
Comment (fee1-dead) [reply src/test/rustdoc/cross-crate-hidden.rs]: revisions do not work - I have added a separate test

**Comment by jyn514** [comment]
@bors r=danielhenrymantilla

Thank you both!

**Pull request merged**